### ISSUE TITLE
Limit places db queries to prevent critical errors

### ISSUE THREAD

**Issue by jaredkerim** [mozilla/activity-stream#1768]
I updated to the latest activity stream on dev, and found that my highlights were not updating.  When I started digging into browser consoles and toolboxes, I found that it was caught in an infinite loop of re requesting metadata, failing, and trying again.  

When I dug a little further, I found that the highlights feed refresh logic was waking up, querying places db for urls, finding 1014 candidate urls, passing those to preview provider, which breaks them into lots of 20 to send to the metadata service, makes 51 concurrent requests to metadata service, attempts to insert 1014 metadata entries into metadatastore, catastrophically fails, and then tries to do the whole thing over again.  

We should limit the number of URLs we receive from placesdb and attempt to send to preview provider.

I've logged some things here to help with debugging, but this caused my activity stream to become unresponsive and I was forced to uninstall.  For reference, I have both local metadata parsing and metadata service enabled.

![screenshot 2016-11-09 15 17 48](https://cloud.githubusercontent.com/assets/119884/20154031/88a0d8ae-a693-11e6-9ac1-5cb2acb0e5c1.png)

**Comment by pdehaan** [comment]
One of the more interesting parts of that screenshot is that we aren't doing templating correctly:

> Failed to fetch metadata by cacheKey: ${e.message}

Looks like we have at least 3 potential violations in /addon/MetadataStore.js:

1. [/addon/MetadataStore.js:448](https://github.com/mozilla/activity-stream/blob/7b7436191542fec546ee3a37914812816ecf0285/addon/MetadataStore.js#L448) should be using backticks for templated strings:
    ```js
Cu.reportError("Failed to fetch metadata by cacheKey: ${e.message}");
```

1. [/addon/MetadataStore.js:471](https://github.com/mozilla/activity-stream/blob/7b7436191542fec546ee3a37914812816ecf0285/addon/MetadataStore.js#L471) should be using backticks for templated strings:
    ```js
Cu.reportError("Failed to fetch metadata by cacheKey: ${e.message}");
```

1. [/addon/MetadataStore.js:492](https://github.com/mozilla/activity-stream/blob/7b7436191542fec546ee3a37914812816ecf0285/addon/MetadataStore.js#L492) should be using backticks for templated strings, and is missing a `$` in `${image.type}`:
    ```js
throw new Error("Fetched unknown image types: {image.type}");
```

**Comment by pdehaan** [comment]
Looks like there is an ESLint rule for this (of course), [`no-template-curly-in-string`](http://eslint.org/docs/rules/no-template-curly-in-string):

```ini
"no-template-curly-in-string": "error",
```

It caught 2-3 of the issues in /addon/MetadataStore.js (but not the third, since it was missing the `$`). But also finds a few false positives in /content-test/, which I'll have to disable errors for:

```js
$ ./node_modules/.bin/eslint .

/Users/pdehaan/dev/github/mozilla/activity-streams/addon/MetadataStore.js
  448:22  error  Unexpected template string expression  no-template-curly-in-string
  471:24  error  Unexpected template string expression  no-template-curly-in-string

/Users/pdehaan/dev/github/mozilla/activity-streams/content-test/lib/getRelativeTime.test.js
   9:6  error  Unexpected template string expression  no-template-curly-in-string
  13:6  error  Unexpected template string expression  no-template-curly-in-string
  17:6  error  Unexpected template string expression  no-template-curly-in-string
  21:6  error  Unexpected template string expression  no-template-curly-in-string

✖ 6 problems (6 errors, 0 warnings)
```

**Comment by ncloudioj** [comment]
Oops, looks like we've gone beyond the limit of binding variables in SQLite (999 by default). Did we change the place query recently? I believe we used to limit that, apparently getting 1000+ links from there sounds incorrect.

**Comment by jaredkerim** [comment]
After some investigation we've discovered that sqlite3 has a hard coded limit to the number of parameters you can send in a single query which is 999.  Because we are querying >999 URLs from places and passing them all in a single query to MetadataStore, we are overflowing the number of parameters we can send in a single query.  Because this will all be smashed to bits and refactored shortly when the great merging occurs, we should for now just cap those queries at 999  results, which according to my testing, still allows activity stream to behave as normal.

See here for more information: 

https://www.sqlite.org/limits.html

Maximum Number Of Host Parameters In A Single SQL Statement

@ncloudioj This is an interesting limitation to sqlite that we should be aware of.

I believe this is also the cause for https://github.com/mozilla/activity-stream/issues/1759

**Comment by jaredkerim** [comment]
This fix should be uplifted to 1.1.7 as it surely affects our SHIELD release.

**Comment by ncloudioj** [comment]
Yup, I was aware of this hard limit in SQLite. I decided not to check this limit in the addon as we used to limit the query result from places.

Anyway, we can add some limit checks so that it could fail gracefully in the metadata store lib.
